core/src/entry_verification_service/worker.rs

```diff
@@ -0,0 +1,524 @@
+        /// next, when updating [`SlotStatus::verified_up_to`].
+        next_data_set: DataSetFirstShred,
+        /// Hash of the very last entry.  Necessary for the PoH verification of the first entry in
+        /// the next data set.
+        last_entry_hash: Hash,
+    },
+}
+
+impl DataSetOptimisticStatus {
+    fn last_entry_hash(&self) -> Option<Hash> {
+        match self {
+            Self::PohStartHash(_) => None,
+            Self::AllButFirst {
+                last_entry_hash, ..
+            } => Some(*last_entry_hash),
+            Self::Verified {
+                last_entry_hash, ..
+            } => Some(*last_entry_hash),
+        }
+    }
+}
+
+struct State {
+    commands: crossbeam_channel::Receiver<Command>,
+
+    blockstore: Arc<Blockstore>,
+
+    processing: HashMap<Slot, SlotStatus>,
+
+    /// Slots that are waiting for the parent slot last entry hash.
+    waiting_for_parent_hash: HashMap<Slot, Vec<Slot>>,
+
+    /// When a client is actively waiting for a certain data set that we did not process yet, we are
+    /// going to skip accumulation in order to provide better latency.
+    ///
+    /// We skip accumulation if we see any work for a data set in this slot, for any of the data
+    /// sets up to and including the specified one.  We also skip accumulation, if we see any work
+    /// for a parent slot.
+    ///
+    /// As slots are relatively big compared to datasets, it is unlikely that the validation service
+    /// will be waiting accumulating data for a grand-parent of the slot the clients are waiting
+    /// for.
+    client_waiting_for: Option<(Slot, DataSetFirstShred)>,
+
+    results: Arc<SharedResults>,
+}
+
+/// Indicates which validation are we planning to do for a data set.
+///
+/// As we are going to run verification for PoH and for signature verification in different threads,
+/// and, potentially, in parallel, holding `entries` as an `Arc<Vec<Entry>>` saves an extra copy.
+#[derive(PartialEq, Eq, Debug, Clone)]
+enum TodoOp {
+    /// Verify PoH for entries, and signature verification for all transactions.
+    All {
+        poh_start_hash: Hash,
+        entries: Arc<Vec<Entry>>,
+    },
+    /// PoH for all but the first entry, and signature verification for all transactions.
+    AllButPohOnFirstEntry { entries: Arc<Vec<Entry>> },
+    /// PoH for the first entry, only.
+    OnlyFirstEntryPoh { poh_start_hash: Hash, entry: Entry },
+}
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+struct TodoItem {
+    /// Slot to mark as failed if the verification specified in `op` fails.
+    slot: Slot,
+    /// Pending verification operation.
+    op: TodoOp,
+}
+
+type Todo = Vec<TodoItem>;
+
+impl State {
+    fn run(&mut self) {
+        loop {
+            let todo = match self.accumulate() {
+                Some(todo) => todo,
+                None => return,
+            };
+
+            self.verify(todo);
+        }
+    }
+
+    fn accumulate(&mut self) -> Option<Todo> {
```

Comment:
Not sure how best to do this, but it would be great to add some metrics so we can gauge the effectiveness of accumulate.

I'm assuming our hope is that by waiting we can avoid the over head of, for example, verifying a dataset except for the first element, then the preceding entry comes in and we loop back to verify the first element (as opposed to just waiting a little bit and then verifying everything in one shot).

Reply:
Absolutely.
I did not write the metrics code yet, but there are a number of things that would be good to measure.
In particular, I wanted to measure the lengths of the `commands` channel.
As it is an indication of how useful the accumulation really is.

As for the wait.  I think, it could actually work forwards and across forks too.
Like we could wait a bit, and possibly, get the next data set, if there is nobody actively waiting for the current data set.
In which case, we can run verification for two data sets at a time.
This could happen if we missed a data set, which would make the replay service wait for a retransmission or a repair.
But the verification service can continue running for subsequent data sets.
It will still run both PoH and signature verification.  It will just skip the very first entry for the PoH hashing.

If not under pressure from the replay service, there might be savings in combining multiple requests together.  From what I read in the verification code (both PoH and signature), there are optimizations to run it in parallel, when enough work is available.
So, I figured, this way, we may also combine completely unrelated data sets together.
For example, when two forks are active, this service will be just fine doing a single GPU invocation for data from both forks.  When they happen to send data sets at approximately the same time.

I also have no idea what would be a good accumulation delay.
Replay service will send `WaitingFor` command when it is actually waiting for something.
So the accumulation delay should not affect the replay service performance [1].
And the only question is: is it accumulating enough to use the efficiency of the parallel verification?
I will need to add metrics to measure the `self.verify` invocation delay.
And the `todo` length.
In addition to the `commands` channel length.
I think measuring averages for these parameters could provide a good understanding if this mechanism is working or not.

[1] I've added a discussion on the `WaitingFor` command potential delays to the enum entry that defines the command:

```rust
    /// There is a chance for this command delivery to be delayed.  If the `commands` channel has
    /// some unfinished work and `todo` has a lot of accumulated work, then we might not start the
    /// verification work for things that the replay service might need us to verify right now.
    ///
    /// Ideally, this should not happen during normal operation.  And should only happen during very
    /// high load, when there is not enough CPU to process all the requests.  In which case, it
    /// seems fine.  Accumulation delay might trigger processing then.
    ///
    /// We need to add metrics to make sure that neither the `commands` channel, nor the `todo` list
    /// do grow too long under normal operation.
    ///
    /// An alternative design would be to send `WaitingFor` though a separate channel.  But it seems
    /// that if it really matters, then the verification is not working fast enough in the first
    /// place.  Also we do not currently track which pending work is needed for `slot`/`data_set` to
    /// be verified, and which work might potentially be delayed for later.  So it is not as simple
    /// as just adding another channel.
```